runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -1961,24 +1960,82 @@ void J9::TransformUtil::removePotentialOSRPointHelperCalls(TR::Compilation* comp
+   TR_ASSERT(comp->supportsInduceOSR() && comp->isOSRTransitionTarget(TR::postExecutionOSR) && comp->getOSRMode() == TR::voluntaryOSR,
+             "prohibitOSROverRange only works in certain modes");
 
    TR::TreeTop* ttAfterEnd = end->getNextTreeTop();
+   TR::TreeTop *tt = start;
 
-   if (comp->getOption(TR_EnableOSR) &&
-       comp->isOSRTransitionTarget(TR::postExecutionOSR) &&
-       comp->getOSRMode() == TR::voluntaryOSR)
+   do
       {
-      TR::TreeTop *tt = start;
-      do
+      TR::Node *osrNode = NULL;
+      if (comp->isPotentialOSRPoint(tt->getNode(), &osrNode))
          {
-         TR::Node *osrNode = NULL;
-         if (comp->isPotentialOSRPoint(tt->getNode(), &osrNode))
-            {
-            dumpOptDetails(comp, "Can no longer OSR at [%p] n%dn\n", osrNode, osrNode->getGlobalIndex());
-            osrNode->getByteCodeInfo().setDoNotProfile(true);
-            }
-         tt = tt->getNextTreeTop();
-         } while (tt != ttAfterEnd);
+         dumpOptDetails(comp, "Can no longer OSR at [%p] n%dn\n", osrNode, osrNode->getGlobalIndex());
+         osrNode->getByteCodeInfo().setDoNotProfile(true);
+         }
+      tt = tt->getNextTreeTop();
+      }
+   while (tt != ttAfterEnd);
+
+   }
+
+static void prohibitGuardedStaticFinalFieldFoldingOnNodeAndChildren(TR::Compilation* comp, TR::Node* node, TR::NodeChecklist *visited)
+   {
+   if (visited->contains(node))
+      return;
+
+   visited->add(node);
+
+   int32_t numChildren = node->getNumChildren();
+   if (numChildren == 0
+       && node->getOpCode().isLoadVarDirect()
+       && node->isLoadOfStaticFinalField()
+       // Field hasn't been folded with known object index
+       && !node->getSymbolReference()->hasKnownObjectIndex())
+      {
+      dumpOptDetails(comp, "Prohibit guarded static final field folding on n%dn %p\n", node->getGlobalIndex(), node);
+      node->getByteCodeInfo().setDoNotProfile(true);
+      return;
       }
+
+   for (int i=0; i<numChildren; i++)
+       {
+       TR::Node* child = node->getChild(i);
+       prohibitGuardedStaticFinalFieldFoldingOnNodeAndChildren(comp, child, visited);
+       }
+   }
+
+/** \brief
+ *     Prohibit guarded static final field folding over range between start and end inclusive.
+ *
+ *     Walk region and ensure all static final field loads has doNotProfile set on their bci.
+ *
+ *  \param comp
+ *     The compilation object.
+ *
+ *  \param start
+ *     The tree marking the start of the range.
+ *
+ *  \param end
+ *     The tree marking the end of the range.
+ */
+void J9::TransformUtil::prohibitGuardedStaticFinalFieldFoldingOverRange(TR::Compilation* comp, TR::TreeTop* start, TR::TreeTop* end)
+   {
+   TR_ASSERT(start->getEnclosingBlock() == end->getEnclosingBlock(), "Does not support range across blocks");
+   TR_ASSERT(comp->supportsInduceOSR() && comp->isOSRTransitionTarget(TR::postExecutionOSR) && comp->getOSRMode() == TR::voluntaryOSR,
```

Comment:
isn't it more correct to say we only do static final field folding in specific modes? the prohibition should work in all modes where we fold?